File: ios-native/Jottery/Jottery/Crypto/KeyManager.swift
import CryptoKit
import Foundation
import LocalAuthentication

/// Manages the in-memory master key, auto-lock timer, and biometric unlock.
///
/// The `SymmetricKey` is kept in memory only — never persisted to disk.
/// Biometric unlock stores the raw key bytes in Keychain with
/// `SecAccessControl` flag `.biometryCurrentSet`.
@MainActor @Observable
final class KeyManager {

    // MARK: - State

    /// The in-memory master key. Internal setter to allow password change flow.
    internal(set) var masterKey: SymmetricKey?
    private(set) var isUnlocked: Bool = false

    /// Seconds since last user interaction before auto-locking.
    var autoLockTimeout: TimeInterval = 15 * 60  // 15 minutes default

    /// Called when the auto-lock timer fires. AppState should wire this
    /// to its own `lock()` so the UI transitions to the unlock screen.
    var onAutoLock: (() -> Void)?

    private var lastActivityDate = Date()
    private var autoLockTimer: Timer?

    // MARK: - Unlock

    /// Derive the master key from a password and store it in memory.
    /// Returns `true` if derivation succeeds.
    func unlock(password: String, salt: Data, iterations: UInt32) -> SymmetricKey {
        let key = CryptoService.deriveKey(password: password, salt: salt, iterations: iterations)
        masterKey = key
        isUnlocked = true
        lastActivityDate = Date()
        startAutoLockTimer()
        return key
    }

    /// Restore the master key from raw bytes (biometric unlock path).
    func unlockWithKeyData(_ keyData: Data) {
        masterKey = SymmetricKey(data: keyData)
        isUnlocked = true
        lastActivityDate = Date()
        startAutoLockTimer()
    }

    /// Lock the application — wipe the in-memory key.
    func lock() {
        masterKey = nil
        isUnlocked = false
        stopAutoLockTimer()
    }

    /// Record user activity to reset the auto-lock countdown.
    func recordActivity() {
        lastActivityDate = Date()
    }

    // MARK: - Biometric Key Storage

    /// Store the current master key in Keychain protected by biometrics.
    func enableBiometricUnlock() throws {
        guard let key = masterKey else {
            throw KeyManagerError.noKeyLoaded
        }
        let keyData = key.withUnsafeBytes { Data($0) }
        try KeychainService.storeBiometricKey(keyData)
    }

    /// Attempt to retrieve the master key via biometric authentication.
    /// Returns `true` if successful.
    func attemptBiometricUnlock() async -> Bool {
        do {
            let keyData = try await KeychainService.retrieveBiometricKey()
            unlockWithKeyData(keyData)
            return true
        } catch {
            return false
        }
    }

    /// Remove biometric key from Keychain.
    func disableBiometricUnlock() {
        KeychainService.deleteBiometricKey()
    }

    /// Check if biometric unlock is available and enrolled.
    var isBiometricAvailable: Bool {
        let context = LAContext()
        var error: NSError?
        return context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
    }

    /// Check if a biometric key is stored.
    var isBiometricEnabled: Bool {
        KeychainService.hasBiometricKey()
    }

    // MARK: - Auto-Lock

    private func startAutoLockTimer() {
        stopAutoLockTimer()
        guard autoLockTimeout > 0 else { return }

        autoLockTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.checkAutoLock()
            }
        }
    }

    private func stopAutoLockTimer() {
        autoLockTimer?.invalidate()
        autoLockTimer = nil
    }

    private func checkAutoLock() {
        guard isUnlocked else { return }
        let elapsed = Date().timeIntervalSince(lastActivityDate)
        if elapsed >= autoLockTimeout {
            // Use the callback so AppState can lock the full app (show unlock screen,
            // clear notes, stop SSE). The callback calls appState.lock() which in turn
            // calls keyManager.lock().
            if let onAutoLock {
                onAutoLock()
            } else {
                lock()
            }
        }
    }

    // No deinit needed — KeyManager lives for the app's lifetime.
    // Timer invalidation happens in lock() / stopAutoLockTimer().
}

enum KeyManagerError: LocalizedError {
    case noKeyLoaded

    var errorDescription: String? {
        switch self {
        case .noKeyLoaded: return "No encryption key loaded"
        }
    }
}

File: android-native/app/src/main/java/com/jottery/android/crypto/KeyManager.kt
package com.jottery.android.crypto

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

/**
 * Manages the in-memory master key, auto-lock timer, and biometric unlock.
 * The SecretKey is kept in memory only — never persisted to disk.
 */
class KeyManager(
    private val scope: CoroutineScope,
) {
    /** The in-memory master key. */
    var masterKey: SecretKey? = null
        internal set

    var isUnlocked: Boolean = false
        private set

    /** Seconds since last user interaction before auto-locking. Default: 15 minutes. */
    var autoLockTimeout: Long = 15 * 60L

    /** Called when the auto-lock timer fires. */
    var onAutoLock: (() -> Unit)? = null

    private var lastActivityTime = System.currentTimeMillis()
    private var autoLockJob: Job? = null

    // MARK: - Unlock

    /** Derive the master key from a password and store it in memory. */
    fun unlock(password: String, salt: ByteArray, iterations: Int): SecretKey {
        val key = CryptoService.deriveKey(password, salt, iterations)
        masterKey = key
        isUnlocked = true
        lastActivityTime = System.currentTimeMillis()
        startAutoLockTimer()
        return key
    }

    /** Restore the master key from raw bytes (biometric unlock path). */
    fun unlockWithKeyData(keyData: ByteArray) {
        masterKey = SecretKeySpec(keyData, "AES")
        isUnlocked = true
        lastActivityTime = System.currentTimeMillis()
        startAutoLockTimer()
    }

    /** Lock the application — wipe the in-memory key. */
    fun lock() {
        masterKey = null
        isUnlocked = false
        stopAutoLockTimer()
    }

    /** Record user activity to reset the auto-lock countdown. */
    fun recordActivity() {
        lastActivityTime = System.currentTimeMillis()
    }

    // MARK: - Auto-Lock

    private fun startAutoLockTimer() {
        stopAutoLockTimer()
        if (autoLockTimeout <= 0) return

        autoLockJob = scope.launch {
            while (true) {
                delay(30_000L) // Check every 30 seconds
                checkAutoLock()
            }
        }
    }

    private fun stopAutoLockTimer() {
        autoLockJob?.cancel()
        autoLockJob = null
    }

    private fun checkAutoLock() {
        if (!isUnlocked) return
        val elapsed = (System.currentTimeMillis() - lastActivityTime) / 1000L
        if (elapsed >= autoLockTimeout) {
            onAutoLock?.invoke() ?: lock()
        }
    }
}
